File: vs-qdrant/node/bench.js
/**
 * Head-to-head harness: official @qdrant/js-client-rest vs nqql (QQL).
 * Same design as python/bench.py — identical scenarios, parity-checked.
 *
 * Usage: node bench.js [--reps 5] [--iters 25]
 * Writes: ../results/node.json
 */
const fs = require("fs");
const path = require("path");
const { performance } = require("perf_hooks");
const { OfficialScenarios } = require("./official_scenarios.js");
const { QqlScenarios } = require("./qql_scenarios.js");

const ROOT = path.resolve(__dirname, "..");
const DATA = path.join(ROOT, "data");
const RESULTS = path.join(ROOT, "results");
const URL = "http://localhost:6333";
const SCROLL_PAGES = 3;
const SCROLL_BATCH = 256;

// ------------------------------------------------------------------ data ----
function loadData() {
  const lines = (f) => fs.readFileSync(path.join(DATA, f), "utf8").split("\n").filter(Boolean);
  const docs = {
    berlin: lines("berlin.jsonl").map((l) => JSON.parse(l)),
    legal: lines("legal.jsonl").map((l) => JSON.parse(l)),
  };
  const f32 = (f, n, d) => {
    const buf = fs.readFileSync(path.join(DATA, f));
    const a = new Float32Array(n * d);
    for (let i = 0; i < n * d; i++) a[i] = buf.readFloatLE(i * 4);
    return a;
  };
  const dense = {
    berlin: f32("berlin_dense.f32", docs.berlin.length, 384),
    legal: f32("legal_dense.f32", docs.legal.length, 384),
  };
  const sparse = {
    berlin: JSON.parse(fs.readFileSync(path.join(DATA, "berlin_bm25.json"), "utf8")),
    legal: JSON.parse(fs.readFileSync(path.join(DATA, "legal_bm25.json"), "utf8")),
  };
  const lens = JSON.parse(fs.readFileSync(path.join(DATA, "legal_colbert_lens.json"), "utf8"));
  const colbertFlatBuf = fs.readFileSync(path.join(DATA, "legal_colbert.f32"));
  const colbertFlat = new Float32Array(lens.reduce((a, b) => a + b, 0) * 128);
  for (let i = 0; i < colbertFlat.length; i++) colbertFlat[i] = colbertFlatBuf.readFloatLE(i * 4);
  // official client ingests nested per-token rows (its PointStruct path is
  // plain JSON arrays); qql side uses the flat Float32Array above.
  const colbertRows = [];
  let o = 0;
  for (const n of lens) {
    const rows = new Array(n);
    for (let t = 0; t < n; t++) {
      const row = new Array(128);
      for (let d = 0; d < 128; d++) row[d] = colbertFlatBuf.readFloatLE((o + t * 128 + d) * 4);
      rows[t] = row;
    }
    colbertRows.push(rows);
    o += n * 128;
  }
  // qql side binds typed arrays (one memcpy); keep a plain-array copy of the
  // sparse terms for the official client's JSON PointStruct path.
  for (const k of ["berlin", "legal"]) {
    sparse[k] = sparse[k].map((sv) => ({
      indicesPlain: sv.indices, valuesPlain: sv.values,
    }));
  }
  const queries = JSON.parse(fs.readFileSync(path.join(DATA, "queries.json"), "utf8"));
  return { docs, dense, sparse, colbertFlat, colbertLens: lens, colbertRows, queries };
}

// --------------------------------------------------------------- timing ----
async function timedRead(fn, reps, iters) {
  const samples = [];
  for (let r = 0; r < reps; r++) {
    if (r === 0) await fn(); // warmup rep untimed
    const t0 = performance.now();
    for (let i = 0; i < iters; i++) await fn();
    samples.push((performance.now() - t0) / iters);
  }
  samples.sort((a, b) => a - b);
  const med = samples[Math.floor(samples.length / 2)];
  return {
    ops_per_sec: Math.round(1000 / med),
    p50_ms: +med.toFixed(3),
    p95_ms: +samples[samples.length - 1].toFixed(3),
    reps, iters,
  };
}

async function waitUntilReady(collection, expected) {
  // Untimed barrier: count visible AND collection green (optimizer idle).
  // Points-count alone races background merges/HNSW indexing, which makes
  // early scenarios slower than later ones on identical data (variance).
  const deadline = Date.now() + 180_000;
  while (Date.now() < deadline) {
    const res = await fetch(`${URL}/collections/${collection}`);
    const info = (await res.json()).result;
    if (info.points_count === expected && info.status === "green") return;
    await new Promise((r) => setTimeout(r, 500));
  }
  throw new Error(`${collection} never reached ${expected} points + green`);
}

// --------------------------------------------------------------- parity ----
function normHits(hits) {
  return hits.map((h) => (typeof h === "object" && !(h instanceof Array) && "id" in h && typeof h.score === "number"
    ? { id: h.id, score: h.score, payload: h.payload }
    : h));
}

function hitKey(h) {
  const n = Number(h.id);
  return Number.isFinite(n) ? n : h.id;
}

function compareHits(official, qql) {
  const o = normHits(official), q = normHits(qql);
  const oIds = o.map(hitKey), qIds = q.map(hitKey);
  const oSet = new Set(oIds), qSet = new Set(qIds);
  let inter = 0;
  for (const id of oSet) if (qSet.has(id)) inter++;
  const overlap = inter / Math.max(1, oSet.size + qSet.size - inter);
  const oScores = new Map(o.map((h) => [hitKey(h), +Number(h.score).toFixed(5)]));
  let maxDiff = 0;
  for (const [k, v] of oScores) {
    const qh = q.find((h) => hitKey(h) === k);
    if (qh) maxDiff = Math.max(maxDiff, Math.abs(v - Number(qh.score)));
  }
  return { top1_match: !!oIds.length && !!qIds.length && oIds[0] === qIds[0],
           jaccard_overlap: +overlap.toFixed(3), max_score_diff: +maxDiff.toFixed(6) };
}

function compareExact(official, qql) {
  const ok = JSON.stringify(official) === JSON.stringify(qql);
  return { match: ok, detail: ok ? "" : `${JSON.stringify(official).slice(0, 100)} != ${JSON.stringify(qql).slice(0, 100)}` };
}

async function restPoint(collection, pointId) {
  const res = await fetch(`${URL}/collections/${collection}/points`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ids: [pointId], with_payload: true, with_vector: true }),
  });
  return (await res.json()).result[0];
}

function arraysClose(a, b, tol = 1e-6) {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (Math.abs(a[i] - b[i]) > tol) return false;
  return true;
}

// ------------------------------------------------------------------ main ----
async function main() {
  const args = { reps: 5, iters: 25 };
  process.argv.forEach((a, i, argv) => {
    if (a === "--reps") args.reps = +argv[i + 1];
    if (a === "--iters") args.iters = +argv[i + 1];
  });

  const { docs, dense, sparse, colbertFlat, colbertLens, colbertRows, queries } = loadData();
  // qql-side sparse terms as typed arrays (Int32Array/Float32Array = one memcpy)
  const sparseQql = {
    berlin: sparse.berlin.map((sv) => ({ indices: new Int32Array(sv.indicesPlain), values: new Float32Array(sv.valuesPlain) })),
    legal: sparse.legal.map((sv) => ({ indices: new Int32Array(sv.indicesPlain), values: new Float32Array(sv.valuesPlain) })),
  };
  // official-side sparse terms as plain arrays (JSON PointStruct path)
  const sparseOff = {
    berlin: sparse.berlin.map((sv) => ({ indices: sv.indicesPlain, values: sv.valuesPlain })),
    legal: sparse.legal.map((sv) => ({ indices: sv.indicesPlain, values: sv.valuesPlain })),
  };
  const official = new OfficialScenarios(URL);
  const qql = new QqlScenarios(URL);
  const R = { meta: await meta(), scenarios: {}, ingest: {}, parity: {}, cold_import: {} };

  // cold import (fresh process, median of 5)
  R.cold_import = {
    "qdrant/js-client-rest": coldRequireMs("@qdrant/js-client-rest"),
    nqql: coldRequireMs("./nqql/index.js"),
  };

  const COLLS = {
    berlin: { official: "vsq_berlin_official", qql: "vsq_berlin_qql" },
    legal: { official: "vsq_legal_official", qql: "vsq_legal_qql" },
  };

  for (const side of ["official", "qql"]) {
    const impl = side === "official" ? official : qql;
    await impl.dropCollection(COLLS.berlin[side]);
    await impl.dropCollection(COLLS.legal[side]);
  }
  await official.createBerlin(COLLS.berlin.official);
  await official.createLegal(COLLS.legal.official);
  await qql.createBerlin(COLLS.berlin.qql);
  await qql.createLegal(COLLS.legal.qql);

  // ingest (timed once per side)
  let t0 = performance.now();
  await official.ingestBerlin(COLLS.berlin.official, docs.berlin, dense.berlin, sparseOff.berlin);
  await official.ingestLegal(COLLS.legal.official, docs.legal, dense.legal, sparseOff.legal, colbertRows);
  R.ingest.official_sec = +((performance.now() - t0) / 1000).toFixed(3);
  t0 = performance.now();
  await qql.ingestBerlin(COLLS.berlin.qql, docs.berlin, dense.berlin, sparseQql.berlin);
  await qql.ingestLegal(COLLS.legal.qql, docs.legal, dense.legal,
    sparseQql.legal, colbertFlat, colbertLens);
  R.ingest.qql_sec = +((performance.now() - t0) / 1000).toFixed(3);
  R.ingest.points = { berlin: docs.berlin.length, legal: docs.legal.length,
                      batch: { berlin: 100, legal: 32 } };
  console.log(`ingest: official ${R.ingest.official_sec}s vs qql ${R.ingest.qql_sec}s`);

  for (const side of ["official", "qql"])
    for (const dom of ["berlin", "legal"])
      await waitUntilReady(COLLS[dom][side], docs[dom].length);

  // vector/payload parity of an ingested point (raw REST, both sides)
  const canon = (o) => JSON.stringify(o, Object.keys(o || {}).sort());
  for (const [dom, pid] of [["berlin", 1], ["legal", 1]]) {
    const o = await restPoint(COLLS[dom].official, pid);
    const q = await restPoint(COLLS[dom].qql, pid);
    let ok = canon(o.payload) === canon(q.payload)
      && new Set(Object.keys(o.vector)).size === new Set(Object.keys(q.vector)).size;
    if (ok) {
      for (const k of Object.keys(o.vector)) {
        const a = o.vector[k], b = q.vector[k];
        if (Array.isArray(a)) ok = ok && arraysClose(a, b);
        else ok = ok && a.indices.join(",") === b.indices.join(",")
          && arraysClose(a.values, b.values);
      }
    }
    R.parity[`ingested_point_${dom}`] = { match: !!ok };
    if (!ok) console.log(`PARITY FAIL ingested_point_${dom}`);
  }

  const S = R.scenarios;
  const B_OFF = COLLS.berlin.official, B_QQL = COLLS.berlin.qql;
  const L_OFF = COLLS.legal.official, L_QQL = COLLS.legal.qql;
  const bq = queries.berlin;

  async function readPair(label, oFn, qFn, parityFn = compareHits) {
    S[label] = {
      official: await timedRead(oFn, args.reps, args.iters),
      qql: await timedRead(qFn, args.reps, args.iters),
    };
    S[label].parity = await parityFn(await oFn(), await qFn());
    const ratio = S[label].qql.ops_per_sec / S[label].official.ops_per_sec;
    console.log(`${label.padEnd(24)} official ${String(S[label].official.ops_per_sec).padStart(9)} | `
      + `qql ${String(S[label].qql.ops_per_sec).padStart(9)} | x${ratio.toFixed(2)} `
      + `| parity ${JSON.stringify(S[label].parity).slice(0, 90)}`);
  }

  await readPair("query_dense",
    () => official.queryDense(B_OFF, bq[0].dense),
    () => qql.queryDense(B_QQL, bq[0].dense));
  await readPair("query_dense_filtered",
    () => official.queryDenseFiltered(B_OFF, bq[1].dense),
    () => qql.queryDenseFiltered(B_QQL, bq[1].dense));
  await readPair("query_sparse",
    () => official.querySparse(B_OFF, bq[0].sparse),
    () => qql.querySparse(B_QQL, bq[0].sparse));
  await readPair("query_hybrid",
    () => official.queryHybrid(B_OFF, bq[2].dense, bq[2].sparse),
    () => qql.queryHybrid(B_QQL, bq[2].dense, bq[2].sparse),
    async (o1, q) => ({
      qql_vs_official: compareHits(o1, q),
      official_vs_official: compareHits(o1,
        await official.queryHybrid(B_OFF, bq[2].dense, bq[2].sparse)),
    }));
  await readPair("scroll_pages",
    () => official.scrollPages(B_OFF, SCROLL_PAGES, SCROLL_BATCH),
    () => qql.scrollPages(B_QQL, SCROLL_PAGES, SCROLL_BATCH),
    (a, b) => compareHits(a.map((id) => ({ id, score: 0 })), b.map((id) => ({ id, score: 0 }))));
  await readPair("count_berlin",
    () => official.countBerlin(B_OFF), () => qql.countBerlin(B_QQL), compareExact);
  await readPair("facet_district",
    () => official.facetDistrict(B_OFF), () => qql.facetDistrict(B_QQL), compareExact);
  await readPair("query_colbert",
    () => official.queryColbert(L_OFF, queries.legal[0].colbert),
    () => qql.queryColbert(L_QQL, queries.legal[0].colbert));
  await readPair("count_legal",
    () => official.countLegal(L_OFF), () => qql.countLegal(L_QQL), compareExact);
  await readPair("prepared_rerun",
    () => official.preparedRerun(B_OFF, bq.map((q) => q.dense)),
    () => qql.preparedRerun(B_QQL, bq.map((q) => q.dense)));

  // writes (last, parity-checked)
  await official.updatePayload(B_OFF);
  await qql.updatePayload(B_QQL);
  R.parity.update_payload = compareExact(
    await official.facetDistrict(B_OFF), await qql.facetDistrict(B_QQL)).match;

  await official.deleteByFilter(B_OFF);
  await qql.deleteByFilter(B_QQL);
  const expected = docs.berlin.filter((d) => d.price <= 250.0).length;
  for (const side of ["official", "qql"])
    await waitUntilReady(COLLS.berlin[side], expected);
  R.parity.delete_by_filter = compareExact(
    await official.countBerlin(B_OFF), await qql.countBerlin(B_QQL)).match;

  fs.mkdirSync(RESULTS, { recursive: true });
  fs.writeFileSync(path.join(RESULTS, "node.json"), JSON.stringify(R, null, 2));
  console.log(`\nwritten ${path.join(RESULTS, "node.json")}`);

  await official.close();
  await qql.close();
}

function CollsName(COLLS, dom, side) {
  return COLLS[dom][side];
}

function coldRequireMs(spec, reps = 5) {
  const { spawnSync } = require("child_process");
  const samples = [];
  for (let i = 0; i < reps; i++) {
    const t0 = performance.now();
    spawnSync(process.execPath, ["-e", `require(${JSON.stringify(spec)})`], {
      cwd: __dirname, stdio: "ignore",
    });
    samples.push(performance.now() - t0);
  }
  samples.sort((a, b) => a - b);
  return +samples[Math.floor(samples.length / 2)].toFixed(1);
}

async function meta() {
  const qdrant = await (await fetch(URL)).json();
  const jsPkg = JSON.parse(fs.readFileSync(
    path.join(__dirname, "node_modules", "@qdrant", "js-client-rest", "package.json"), "utf8"));
  return {
    timestamp: new Date().toISOString(),
    node: process.version,
    nqql: require("./nqql/index.js").version,
    js_client_rest: jsPkg.version,
    qdrant_server: qdrant.version,
    url: URL,
  };
}

main().catch((e) => { console.error(e); process.exit(1); });

File: vs-qdrant/python/bench.py
#!/usr/bin/env python3
"""Head-to-head harness: official qdrant-client vs pyqql (QQL), live Qdrant.

Design (identical across the python/node/rust legs):
  * Both contenders build their OWN collection from byte-identical data
    (data/ artifacts, checksummed in manifest.json).
  * Ingest: each side's idiomatic batched upsert path, wait=false on both,
    wall-clock timed. A durability barrier (untimed) precedes all reads.
  * Reads: warmup + median-of-reps ops/s with p50/p95 latency.
  * Parity: every read scenario compares results between the two collections
    (ids, scores, facet maps, counts). Sparse/facet/count must match exactly;
    ANN scenarios must match on top-1 with >=80% top-10 overlap.
  * Writes (update payload, delete-by-filter) run once, last, with parity.

Usage: python3 bench.py [--reps 5] [--iters 25]
Writes: ../results/python.json
"""
from __future__ import annotations

import argparse
import json
import statistics
import subprocess
import sys
import time
from pathlib import Path

import numpy as np

HERE = Path(__file__).resolve().parent
ROOT = HERE.parent
sys.path.insert(0, str(HERE))

from config import BATCH_BERLIN, BATCH_LEGAL, COLLS, LIMIT, SCROLL_BATCH, SCROLL_PAGES, URL  # noqa: E402

from official_scenarios import OfficialScenarios  # noqa: E402
from qql_scenarios import QqlScenarios  # noqa: E402

DATA = ROOT / "data"
RESULTS = ROOT / "results"


# ------------------------------------------------------------------ data ----
def load_data():
    def f32(name, shape):
        a = np.fromfile(DATA / name, dtype=np.float32)
        return a.reshape(shape)

    docs = {k: [json.loads(l) for l in (DATA / f"{k}.jsonl").read_text().splitlines() if l]
            for k in ("berlin", "legal")}
    dense = {
        "berlin": f32("berlin_dense.f32", (len(docs["berlin"]), 384)),
        "legal": f32("legal_dense.f32", (len(docs["legal"]), 384)),
    }
    sparse = {k: json.loads((DATA / f"{k}_bm25.json").read_text()) for k in ("berlin", "legal")}
    # tf values as f32 numpy: bind via the memoryview fast path (one copy)
    for k in sparse:
        for sv in sparse[k]:
            sv["values"] = np.asarray(sv["values"], dtype=np.float32)
    lens = json.loads((DATA / "legal_colbert_lens.json").read_text())
    colbert_flat = np.fromfile(DATA / "legal_colbert.f32", dtype=np.float32)
    queries = json.loads((DATA / "queries.json").read_text())
    return docs, dense, sparse, colbert_flat, lens, queries


# --------------------------------------------------------------- timing ----
def timed_read(fn, reps: int, iters: int) -> dict:
    """Median ops/s + latency percentiles over reps runs of `iters` calls."""
    samples_ms = []
    for r in range(reps):
        fn() if r == 0 else None                      # warmup rep untimed
        t0 = time.perf_counter()
        for _ in range(iters):
            fn()
        dt = (time.perf_counter() - t0) / iters
        samples_ms.append(dt * 1000.0)
    med = statistics.median(samples_ms)
    return {"ops_per_sec": round(1000.0 / med), "p50_ms": round(med, 3),
            "p95_ms": round(max(samples_ms), 3), "reps": reps, "iters": iters}


def wait_until_ready(client_url: str, collection: str, expected: int) -> None:
    """Untimed barrier: block until the count is visible AND the collection
    is green (optimizer idle). Points-count alone races background segment
    merges/HNSW indexing, which makes early read scenarios slower than later
    ones on the same data — pure run-to-run variance, not engine signal."""
    import urllib.request
    deadline = time.time() + 180
    while time.time() < deadline:
        with urllib.request.urlopen(f"{client_url}/collections/{collection}") as r:
            info = json.loads(r.read())["result"]
        if info.get("points_count") == expected and info.get("status") == "green":
            return
        time.sleep(0.5)
    raise RuntimeError(f"{collection} never reached {expected} points + green")


# --------------------------------------------------------------- parity ----
def norm_hits(hits) -> list[dict]:
    """Official SDK returns ScoredPoint objects; QQL returns dicts."""
    return [h if isinstance(h, dict) else
            {"id": h.id, "score": h.score, "payload": h.payload} for h in hits]


def hit_key(h) -> tuple:
    """QQL may return ids as strings ("1"), the official SDK as ints."""
    try:
        return int(h["id"])
    except (ValueError, TypeError):
        return h["id"]


def rest_point(url: str, collection: str, point_id: int) -> dict:
    """Raw-REST point fetch (harness-level parity probe, not a timed path)."""
    import urllib.request
    body = json.dumps({"ids": [point_id], "with_payload": True, "with_vector": True}).encode()
    req = urllib.request.Request(
        f"{url}/collections/{collection}/points", data=body,
        headers={"Content-Type": "application/json"}, method="POST")
    with urllib.request.urlopen(req) as r:
        return json.loads(r.read())["result"][0]


def compare_hits(official: list, qql: list) -> dict:
    official, qql = norm_hits(official), norm_hits(qql)
    o_ids = [hit_key(h) for h in official]
    q_ids = [hit_key(h) for h in qql]
    o_set, q_set = set(o_ids), set(q_ids)
    overlap = len(o_set & q_set) / max(1, len(o_set | q_set))
    top1 = bool(o_ids and q_ids and o_ids[0] == q_ids[0])
    o_scores = {hit_key(h): round(float(h["score"]), 5) for h in official}
    q_scores = {hit_key(h): round(float(h["score"]), 5) for h in qql}
    diffs = [abs(o_scores[k] - q_scores[k]) for k in o_scores.keys() & q_scores.keys()]
    return {"top1_match": top1, "jaccard_overlap": round(overlap, 3),
            "max_score_diff": max(diffs) if diffs else 0.0}


def compare_exact(official, qql) -> dict:
    ok = official == qql
    return {"match": ok, "detail": "" if ok else f"{official!r} != {qql!r}"[:200]}


# ------------------------------------------------------------------ main ----
def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--reps", type=int, default=5)
    ap.add_argument("--iters", type=int, default=25)
    args = ap.parse_args()

    docs, dense, sparse, colbert_flat, colbert_lens, queries = load_data()
    official, qql = OfficialScenarios(URL), QqlScenarios(URL)
    R = {"meta": meta(), "scenarios": {}, "ingest": {}, "parity": {}, "cold_import": {}}

    # ---- cold import (fresh interpreter, median of N) ----
    R["cold_import"] = {
        "qdrant_client": cold_import_ms("qdrant_client"),
        "pyqql": cold_import_ms("pyqql"),
    }

    # ---- setup: drop, create ----
    for side_name, side in (("official", official), ("qql", qql)):
        for dom in ("berlin", "legal"):
            side.drop_collection(COLLS[dom][side_name])
    official.create_berlin(COLLS["berlin"]["official"])
    official.create_legal(COLLS["legal"]["official"])
    qql.create_berlin(COLLS["berlin"]["qql"])
    qql.create_legal(COLLS["legal"]["qql"])

    # ---- ingest (timed once per side) ----
    t0 = time.perf_counter()
    official.ingest_berlin(COLLS["berlin"]["official"], docs["berlin"], dense["berlin"], sparse["berlin"])
    colbert_rows, o = [], 0
    for n in colbert_lens:
        colbert_rows.append(colbert_flat[o:o + n * 128].reshape(n, 128).tolist())
        o += n * 128
    official.ingest_legal(COLLS["legal"]["official"], docs["legal"], dense["legal"], sparse["legal"], colbert_rows)
    R["ingest"]["official_sec"] = round(time.perf_counter() - t0, 3)
    t0 = time.perf_counter()
    qql.ingest_berlin(COLLS["berlin"]["qql"], docs["berlin"], dense["berlin"], sparse["berlin"])
    qql.ingest_legal(COLLS["legal"]["qql"], docs["legal"], dense["legal"], sparse["legal"], colbert_flat, colbert_lens)
    R["ingest"]["qql_sec"] = round(time.perf_counter() - t0, 3)
    R["ingest"]["points"] = {"berlin": len(docs["berlin"]), "legal": len(docs["legal"]),
                             "batch": {"berlin": BATCH_BERLIN, "legal": BATCH_LEGAL}}
    print(f"ingest: official {R['ingest']['official_sec']}s vs qql {R['ingest']['qql_sec']}s", flush=True)

    # durability barrier (untimed)
    for side_name in ("official", "qql"):
        for dom in ("berlin", "legal"):
            wait_until_ready(URL, COLLS[dom][side_name], len(docs[dom]))

    # ---- vector/payload parity of an ingested point (raw REST, both sides) ----
    for dom, pid in (("berlin", 1), ("legal", 1)):
        o = rest_point(URL, COLLS[dom]["official"], pid)
        q = rest_point(URL, COLLS[dom]["qql"], pid)
        ok = (o["payload"] == q["payload"]
              and o["vector"].keys() == q["vector"].keys()
              and all(np.allclose(o["vector"][k], q["vector"][k], atol=1e-6)
                      if isinstance(o["vector"][k], list)
                      else (o["vector"][k]["indices"] == q["vector"][k]["indices"]
                            and np.allclose(o["vector"][k]["values"],
                                            q["vector"][k]["values"], atol=1e-6))
                      for k in o["vector"]))
        R["parity"][f"ingested_point_{dom}"] = {"match": bool(ok)}
        if not ok:
            print(f"PARITY FAIL ingested_point_{dom}", flush=True)

    # ---- read scenarios ----
    S = R["scenarios"]
    B_OFF, B_QQL = COLLS["berlin"]["official"], COLLS["berlin"]["qql"]
    L_OFF, L_QQL = COLLS["legal"]["official"], COLLS["legal"]["qql"]
    bq = queries["berlin"]

    def read_pair(label, o_fn, q_fn, parity_fn=compare_hits, require_exact=False):
        S[label] = {"official": timed_read(lambda: o_fn(), args.reps, args.iters),
                    "qql": timed_read(lambda: q_fn(), args.reps, args.iters)}
        S[label]["parity"] = parity_fn(o_fn(), q_fn())
        ratio = S[label]["qql"]["ops_per_sec"] / S[label]["official"]["ops_per_sec"]
        print(f"{label:24s} official {S[label]['official']['ops_per_sec']:>9,} | "
              f"qql {S[label]['qql']['ops_per_sec']:>9,} | x{ratio:.2f} "
              f"| parity {json.dumps(S[label]['parity'])[:90]}", flush=True)

    read_pair("query_dense",
              lambda: official.query_dense(B_OFF, bq[0]["dense"]),
              lambda: qql.query_dense(B_QQL, bq[0]["dense"]))
    read_pair("query_dense_filtered",
              lambda: official.query_dense_filtered(B_OFF, bq[1]["dense"]),
              lambda: qql.query_dense_filtered(B_QQL, bq[1]["dense"]))
    read_pair("query_sparse",
              lambda: official.query_sparse(B_OFF, bq[0]["sparse"]),
              lambda: qql.query_sparse(B_QQL, bq[0]["sparse"]))
    # hybrid: RRF ordering is nondeterministic at equal-score ties even
    # within one SDK (server-side tie-breaking) — baseline official-vs-official
    # self-consistency so the qql-vs-official delta has a fair reference.
    read_pair("query_hybrid",
              lambda: official.query_hybrid(B_OFF, bq[2]["dense"], bq[2]["sparse"]),
              lambda: qql.query_hybrid(B_QQL, bq[2]["dense"], bq[2]["sparse"]),
              parity_fn=lambda o1, q: {
                  "qql_vs_official": compare_hits(o1, q),
                  "official_vs_official": compare_hits(
                      o1, official.query_hybrid(B_OFF, bq[2]["dense"], bq[2]["sparse"])),
              })
    read_pair("scroll_pages",
              lambda: official.scroll_pages(B_OFF, SCROLL_PAGES, SCROLL_BATCH),
              lambda: qql.scroll_pages(B_QQL, SCROLL_PAGES, SCROLL_BATCH),
              parity_fn=lambda a, b: compare_hits(
                  [{"id": i, "score": 0} for i in a], [{"id": i, "score": 0} for i in b]))
    read_pair("count_berlin",
              lambda: official.count_berlin(B_OFF),
              lambda: qql.count_berlin(B_QQL), parity_fn=compare_exact)
    read_pair("facet_district",
              lambda: official.facet_district(B_OFF),
              lambda: qql.facet_district(B_QQL), parity_fn=compare_exact)
    read_pair("query_colbert",
              lambda: official.query_colbert(L_OFF, queries["legal"][0]["colbert"]),
              lambda: qql.query_colbert(L_QQL, queries["legal"][0]["colbert"]))
    read_pair("count_legal",
              lambda: official.count_legal(L_OFF),
              lambda: qql.count_legal(L_QQL), parity_fn=compare_exact)

    # ---- prepared rerun (many small parameterized calls) ----
    read_pair("prepared_rerun",
              lambda: official.prepared_rerun(B_OFF, [q["dense"] for q in bq]),
              lambda: qql.prepared_rerun(B_QQL, [q["dense"] for q in bq]))

    # ---- writes (last, parity-checked) ----
    official.update_payload(B_OFF)
    qql.update_payload(B_QQL)
    R["parity"]["update_payload"] = compare_exact(
        official.facet_district(B_OFF), qql.facet_district(B_QQL))["match"]

    official.delete_by_filter(B_OFF)
    qql.delete_by_filter(B_QQL)
    for side_name in ("official", "qql"):
        wait_until_ready(URL, COLLS["berlin"][side_name],
                         sum(1 for d in docs["berlin"] if d["price"] <= 250.0))
    R["parity"]["delete_by_filter"] = compare_exact(
        official.count_berlin(B_OFF), qql.count_berlin(B_QQL))["match"]

    RESULTS.mkdir(exist_ok=True)
    out = RESULTS / "python.json"
    out.write_text(json.dumps(R, indent=2))
    print(f"\nwritten {out}")

    official.close()
    qql.close()


def cold_import_ms(module: str, reps: int = 5) -> float:
    samples = []
    for _ in range(reps):
        t0 = time.perf_counter()
        subprocess.run([sys.executable, "-c", f"import {module}"], check=True, capture_output=True)
        samples.append((time.perf_counter() - t0) * 1000)
    return round(statistics.median(samples), 1)


def meta() -> dict:
    import importlib.metadata
    import platform
    import pyqql
    import qdrant_client
    import urllib.request
    with urllib.request.urlopen("http://localhost:6333") as r:
        server = json.loads(r.read())
    return {
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S"),
        "python": platform.python_version(),
        "pyqql": pyqql.__version__,
        "qdrant_client": importlib.metadata.version("qdrant-client"),
        "qdrant_server": server["version"],
        "url": URL,
    }


if __name__ == "__main__":
    main()
